src/bin/server.rs

```diff
@@ -38,26 +39,71 @@ fn main() {
     cfg.set_str("user.name", "bors").unwrap();
     cfg.set_str("user.email", "bors@rust-lang.org").unwrap();
 
+    let api_protocol = String::from("https");
+    let mirror = env::var("MIRROR").is_ok();
+
     let heroku = env::var("HEROKU").is_ok();
     let cargo_env = if heroku {
         cargo_registry::Env::Production
     } else {
         cargo_registry::Env::Development
     };
+
+    let uploader = match (cargo_env, mirror) {
+        (cargo_registry::Env::Production, false) => {
+            // `env` panics if these vars are not set
+            cargo_registry::Uploader::S3 {
+                bucket: s3::Bucket::new(env("S3_BUCKET"),
+                                        env::var("S3_REGION").ok(),
+                                        env("S3_ACCESS_KEY"),
+                                        env("S3_SECRET_KEY"),
+                                        &api_protocol),
+                proxy: None,
+            }
+        },
+        (cargo_registry::Env::Production, true) => {
+            // Read-only mirrors don't need access key or secret key,
+            // but they might have them. Definitely need bucket though.
+            cargo_registry::Uploader::S3 {
+                bucket: s3::Bucket::new(env("S3_BUCKET"),
+                                        env::var("S3_REGION").ok(),
+                                        env::var("S3_ACCESS_KEY").unwrap_or(String::new()),
+                                        env::var("S3_SECRET_KEY").unwrap_or(String::new()),
+                                        &api_protocol),
+                proxy: None,
+            }
+        },
+        (cargo_registry::Env::Development, _) => {
+            if env::var("S3_BUCKET").is_ok() {
+                println!("Using S3 uploader");
+                cargo_registry::Uploader::S3 {
+                    bucket: s3::Bucket::new(env("S3_BUCKET"),
+                                            env::var("S3_REGION").ok(),
+                                            env::var("S3_ACCESS_KEY").unwrap_or(String::new()),
+                                            env::var("S3_SECRET_KEY").unwrap_or(String::new()),
+                                            &api_protocol),
+                    proxy: None,
+                }
+            } else {
+                println!("Using local uploader, crate files will be in the dist directory");
+                cargo_registry::Uploader::Local
+            }
+        },
+        // See immediately before this match where we choose either prod or dev
+        (cargo_registry::Env::Test, _) => unreachable!(),
```

Comment:
Since we only care about "prod" and "not prod", do you think it makes sense to have those be the match arms?